📖 Add documentation around KubeadmControlPlane (#2607)
kubernetes-sigs/cluster-api: docs/book/src/tasks/kubeadm-control-plane.md

> # Kubeadm control plane

Reviewer: This doc currently seems to be more around just the upgrade process. I'm wondering if it might be better to title it 'Upgrading workload clusters' and have a short blurb that the use of KubeadmControlPlane and MachineDeployment is a requirement for supported upgrades.

Author: The intent is that this is a place to put tasks related to kubeadm control plane. Upgrades just happen to be the most interesting feature. There should probably be some sentences about scaling up and down, fixing a busted control plane node and other tasks that kubeadm control plane allows us to do

Using the Kubeadm control plane type to manage a control plane provides several ways to upgrade control plane machines.

## Upgrading

The high level steps to fully upgrading a workload cluster are to first upgrade the control plane and then upgrade
the worker machines.

> ### Upgrading workload machines managed by a `MachineDeployment`

Reviewer: Would it make sense to move this section to the end, since we are telling users that upgrading a MachineDeployment should be done last?

Reviewer 2: +1

> It is recommended to manage workload machines with a `MachineDeployment` as this will transparently manage `MachineSet`s

Reviewer: ```suggestion
It is recommended to manage workload machines with one ore more `MachineDeployment`s as this will transparently manage `MachineSet`s
```

> and `Machine`s to allow for a seamless scaling experience. A modification to the `MachineDeployment`s spec will begin a

Reviewer: I can't come up with better wording at the moment, but I think it might be better to give a better description of how the MachineDeployment is orchestrating the rollout of the change (and allowing for the ability to rollback in the event of an issue)

rolling update of the workload machines.

> ### How to upgrade the underlying machine image

Reviewer 2: What about grouping this paragraph and the following under a section:
Upgrading control-plane machines managed by a `KubeadmControlPlane` object

To upgrade the control plane machines underlying machine images, the `MachineTemplate` resource referenced by the
`KubeadmControlPlane` must be changed. Since `MachineTemplate` resources are immutable, the recommended approach is to

> copy the existing `MachineTemplate`, modify the necessary values, create the new `MachineTemplate` on the management

Reviewer 2: "modify the necessary values" might confuse the reader, is it possible to link somehow this sentence with the next paragraph (where there is the explanation of what should be modified)
probably a simple "(see next paragraph for an explanation of the possible changes)" might suffice...

cluster, and finally, modify the existing `KubeadmControlPlane` resource to reference the new `MachineTemplate`
resource. This will trigger a rolling update of the control plane using the new values found in the `MachineTemplate`.

### How to upgrade the Kubernetes control plane version

To upgrade the Kubernetes control plane version, which will likely, depending on the provider, also upgrade the
underlying machine image, make a modification to the `KubeadmControlPlane` resource's `Version` field. This will trigger

> a rolling upgrade of the control plane.

Reviewer: Might be good to add a statement along the lines of:

```
If the infrastructure provider used requires updating image IDs to update the Kubernetes version, the update for the MachineTemplate must be done at the same time as updating the KubeadmControlPlane.Spec.Version field to the appropriate version for the image. This is to ensure that the deployed control plane pods and additional upgrade steps, such as updating the kube-proxy daemonset for the workload cluster, will use the correct matching Kubernetes version as the on-disk binaries provided by the image.
```

Might be good to add a similar blurb related to image based overrides for the provided KubeadmConfig, but both of these can easily be done later and shouldn't necessarily be blocking for this PR.

Author: I added a paragraph around the infrastructure provider changes but not a section for the KubeadmConfig. 

Let's add that along with how to upgrade etcd and add ons maybe after we get this in place?

Using the Kubeadm control plane type to manage a control plane provides several ways to upgrade control plane machines.

## Upgrading workload clusters

The high level steps to fully upgrading a workload cluster are to first upgrade the control plane and then upgrade
the worker machines.

### Upgrading the control plane machines

#### How to upgrade the underlying machine image

To upgrade the control plane machines underlying machine images, the `MachineTemplate` resource referenced by the
`KubeadmControlPlane` must be changed. Since `MachineTemplate` resources are immutable, the recommended approach is to

1. Copy the existing `MachineTemplate`.
2. Modify the values that need changing, such as instance type or image ID.
3. Create the new `MachineTemplate` on the management cluster.
4. Modify the existing `KubeadmControlPlane` resource to reference the new `MachineTemplate` resource.

The final step will trigger a rolling update of the control plane using the new values found in the `MachineTemplate`.

#### How to upgrade the Kubernetes control plane version

To upgrade the Kubernetes control plane version, which will likely, depending on the provider, also upgrade the

> underlying machine image, make a modification to the `KubeadmControlPlane` resource's `Version` field. This will trigger

Reviewer: nit:
```suggestion
underlying machine image, make a modification to the `KubeadmControlPlane` resource's `Spec.Version` field. This will trigger
```

a rolling upgrade of the control plane.

Some infrastructure providers, such as [CAPA](https://github.com/kubernetes-sigs/cluster-api-provider-aws), require

> the machine image to match the Kubernetes version specified in the `KubeadmControlPlane` spec. In order to only trigger

Reviewer: ```suggestion
that if a specific machine image is specified it has to match the Kubernetes version specified in the `KubeadmControlPlane` spec. In order to only trigger
```

Tried to see if I could add some clarity that it's only a requirement for CAPA if the image is specified vs the default lookup.
